compiler/rustc_codegen_ssa/src/back/metadata.rs

```diff
@@ -205,6 +206,48 @@ pub(crate) fn create_object_file(sess: &Session) -> Option<write::Object<'static
         _ => elf::ELFOSABI_NONE,
     };
     let abi_version = 0;
+    // check bti protection
+    let mut check_cfprotection = false;
+    let mut pr_type: u32 = 0;
+    if architecture == Architecture::X86_64
+        && matches!(
+            sess.opts.unstable_opts.cf_protection,
+            rustc_session::config::CFProtection::Branch | rustc_session::config::CFProtection::Full
+        )
+    {
+        check_cfprotection = true;
+        pr_type = 0xc0000002;
+    }
+    if architecture == Architecture::Aarch64 && sess.opts.unstable_opts.branch_protection.is_some()
+    {
+        check_cfprotection = true;
+        pr_type = 0xc0000000;
+    }
+    if check_cfprotection && binary_format == BinaryFormat::Elf {
+        let segment: Vec<u8> = file.segment_name(StandardSegment::Data).to_vec();
+        let name: Vec<u8> = b".note.gnu.property".to_vec();
+        let kind = SectionKind::Note;
+        let section = file.add_section(segment, name, kind);
+        let mut data: Vec<u8> = Vec::new();
+        let n_namsz: u32 = 4; // Size of the n_name field
+        let n_descsz: u32 = 16; // Size of the n_desc field
+        let n_type: u32 = NT_GNU_PROPERTY_TYPE_0; // Type of note descriptor
+        let values = [n_namsz, n_descsz, n_type];
+        values.map(|v| data.extend_from_slice(&(v.to_ne_bytes())));
+        data.push(b'G'); // Owner of the program property note
+        data.push(b'N');
+        data.push(b'U');
+        data.push(0);
+        let pr_datasz: u32 = 4; //size of the pr_data field
+        let pr_data: u32 = 3; //program property descriptor
+        let pr_padding: u32 = 3;
+        let values = [pr_type, pr_datasz, pr_data, pr_padding];
+        values.map(|v| data.extend_from_slice(&(v.to_ne_bytes())));
+        let x = data.len();
+        assert_eq!(x, 32);
+        let align = 4;
+        file.append_section_data(section, &data, align);
+    }
```

Comment:
I think it would be slightly tidier if the logic here was split out into its own function and just invoked here.

Reply:
ok! done.